File: app/src/main/java/com/example/babajidemustapha/survey/shared/utils/DbOperationHelper.java
package com.example.babajidemustapha.survey.shared.utils;

import android.os.AsyncTask;

public class DbOperationHelper<Result> extends AsyncTask<Void, Void, Result> {

    private IDbOperationHelper<Result> dbOperationHelper;

    private DbOperationHelper(IDbOperationHelper<Result> dbOperationHelper) {
        this.dbOperationHelper = dbOperationHelper;
    }

    public static <T> void execute(IDbOperationHelper<T> dbOperationHelper) {
        new DbOperationHelper<>(dbOperationHelper).execute();
    }

    @Override
    protected Result doInBackground(Void... voids) {
        return dbOperationHelper.run();
    }

    @Override
    protected void onPostExecute(Result result) {
        super.onPostExecute(result);
        dbOperationHelper.onCompleted(result);
    }

    public interface IDbOperationHelper<TResult> {
        TResult run();

        void onCompleted(TResult result);
    }
}

File: app/src/main/java/com/example/babajidemustapha/survey/shared/utils/DbOperationHelper.kt
package com.example.babajidemustapha.survey.shared.utils

import android.os.AsyncTask

class DbOperationHelper<Result> private constructor(private val dbOperationHelper: IDbOperationHelper<Result>) : AsyncTask<Void?, Void?, Result>() {

    override fun onPostExecute(result: Result) {
        super.onPostExecute(result)
        dbOperationHelper.onCompleted(result)
    }

    interface IDbOperationHelper<TResult> {
        fun run(): TResult
        fun onCompleted(result: TResult)
    }

    companion object {
        fun <T> execute(dbOperationHelper: IDbOperationHelper<T>) {
            DbOperationHelper(dbOperationHelper).execute()
        }
    }

    override fun doInBackground(vararg params: Void?): Result {
        return dbOperationHelper.run()
    }
}